src/toxcore/tcp/server/server.rs

```diff
@@ -365,6 +368,41 @@ impl Server {
             Box::new( future::ok(()) )
         }
     }
+    /* Remove timedout connected clients
+    */
+    fn remove_timedout_clients(&self) -> IoFuture<()> {
+        let keys = self.state.read().connected_clients.iter()
+            .filter(|(_key, client)| client.is_pong_timedout())
+            .map(|(key, _client)| *key)
+            .collect::<Vec<PublicKey>>();
+
+        let remove_timedouts = keys.iter()
+            .map(|key| {
+                self.shutdown_client(key)
```

Comment:
Every call to this method acquires new lock which is not good. Besides performance impact it may lead to error when we try to remove already removed client (if it was removed after we collected keys). I'd prefer using single lock and creating inner `shutdown_client` function that takes state as a parameter instead of taking it from `RwLock` so that we can acquire this lock only once inside `send_pings`. What do you think?

Reply:
Good idea.